Keycloak Custom User Provider

jacob6838 opened the pull request:
<!--- Thanks for the contribution! -->

# PR Details

Note: This is a Draft PR, because updates still need to be made to the cvmanager python api to account for small changes to the public.users table. 
There will be future PRs to begin to utilize the jwt org/role information in the api and webapp.

## Description

Adding a custom user provider java project to the CVManager project. This is configured to generate a .jar file, which is picked up by the Dockerfile and Realm.json. 

This custom user provider enables Keycloak to query and update the public.users table, and view organization/role data from the public.user_organization, public.organizations, and public.roles tables.

This java project also include a custom token mapper, which incorporates user data and organizations+roles into the JWT token like this:
...
"cvmanager_data": {
    "super_user": "1",
    "organizations": [
        {
            "org": "Test Org",
            "role": "admin"
        },
        {
            "org": "Test Org 2",
            "role": "user"
        }
    ],
    "user_created_timestamp": 1719549350683
},
...

## How Has This Been Tested?

This was tested through building the keycloak image, and going through each of the expected user viewing and management actions through the keycloak admin console, as well as generating tokens through the admin REST api.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] My changes require new environment variables:
  - [x] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [x] My changes require updates to the documentation:
  - [x] I have updated the documentation accordingly.
- [ ] My changes require updates and/or additions to the unit tests:
  - [ ] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

jacob6838 committed:
formatting realm.json

jacob6838 committed:
Adding custom postgres-user-provider

jacob6838 committed:
Update Dockerfile

jacob6838 committed:
Adding keycloak custom-user-provider java project

jacob6838 committed:
Fixing custom-user-provider

jacob6838 committed:
Update README.md

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider

jacob6838 committed:
Adding default uuid generator for user_id

John-Wiens committed:
Adding automatic docker builder for keycloak-custom-providers

jacob6838 committed:
Merge pull request #87 from CDOT-CV/keycloak-docker-builder

Adding automatic docker builder for keycloak-custom-providers

jacob6838 committed:
Removing error_emails from custom user provider

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider

jacob6838 committed:
Adding keycloak_id and user_id

jacob6838 committed:
Updating logging

jacob6838 committed:
updating documentation

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider

jacob6838 committed:
updating keycloak health checks

jacob6838 committed:
Fixing keycloak health checks

payneBrandon commented:
@jacob6838 looks like a merge conflict after the last PR merged in

payneBrandon commented on resources/keycloak/realm.json:
```diff
@@ -5150,28 +5150,33 @@
-        "realmRoles": ["ADMIN"],
-        "clientRoles": {
-          "realm-management": ["manage-users", "realm-admin"]
+        "id": "f:a6d0ac0a-279e-40f6-a226-81cc7d5be501:e3f4884b-7cb4-46cc-a353-1314ac2ff03e",
```
What are these two federated users?

jacob6838 replied:
These are two entries which add the keycloak-side of the user's data, since keycloak no longer stores the base user information. The keycloak-side data includes the email_verified attribute and similar, as well as the user's password (credential block). Then, the rest of the user's base info (user_id, keycloak_id, name, ...) is all in postgres

payneBrandon replied:
So this is just relating to the email_verified attribute and credential information at a high level and not a particular user? The way this read with being under "federatedUsers" it makes it look like we're adding in two explicit users.

jacob6838 replied:
I don't really understand why they are different entries, but they both apply to the same user (you can see the first half of the id is the same). The first half of the id corresponds to the user id injected into postgres, which is how they are applied

payneBrandon commented on resources/keycloak/Dockerfile:
```diff
@@ -4,12 +4,37 @@ RUN dnf install --installroot /mnt/rootfs curl jq \
+COPY --from=jarbuilder /home/custom-user-provider/target/keycloak-custom-providers.jar /opt/keycloak/providers/
+
+# for demonstration purposes only, please make sure to use proper certificates in production instead
+RUN keytool -genkeypair -storepass password -storetype PKCS12 -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore conf/server.keystore
```
What's our plan to handle this?

jacob6838 replied:
I just created a new dockerfile for local dev named Dockerfile_localdev, which uses a self-signed cert. I also added some documention to the keycloak folder, with instructions about including a certificate in the production Dockerfile image

payneBrandon replied:
I'm not seeing a dockerfile_localdev version, but I do see the Dockerfile has been updated to pull in generic named certs. Is this what you were talking about?

jacob6838 replied:
Sorry this got overwritten by a second change. I added a dev script to generate a local self-signed cert, and that local cert gets pulled into the Dockerfile

payneBrandon commented on resources/keycloak/custom-user-provider/README.md:
````diff
@@ -0,0 +1,98 @@
+...
+```
+
+Keycloak custom toke mapper walkthrough: [Custom Protocol Mapper with Keycloak](https://www.baeldung.com/keycloak-custom-protocol-mapper)
````
It looks like we've copy/pasted some of the Baeldung classes in here and kept the package using their nomenclature (com.baeldung.xxx). Is this intended/required since we used their sytems to build from? Should we consider changing that to match our project?

jacob6838 replied:
I updated the application name/paths!

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider

payneBrandon commented on resources/keycloak/custom-user-provider/src/main/java/com/baeldung/auth/provider/mapper/CustomProtocolMapper.java:
```diff
@@ -0,0 +1,74 @@
+                    userSession.getUser().getId());
+
+            // Map custom property to the access token
+            transformAccessToken.getOtherClaims().put("cvmanager_data", UserObject.toMap(user));
```
Are we just mapping the whole user object to this? Is there not extra properties in there we may not want? Let's add in some comments explaining why we're adding this.

jacob6838 replied:
The toMap is a custom method, I renamed it to toTokenMap and added some comments

payneBrandon commented on resources/keycloak/custom-user-provider/insert.sql:
```diff
@@ -0,0 +1,5 @@
+INSERT INTO users ("username", "firstName", "lastName", "email", "birthDate")
```
Is this referenced anywhere or just included as a demo?

jacob6838 replied:
That was a test script, removed!

payneBrandon commented on resources/keycloak/custom-user-provider/target/maven-status/maven-compiler-plugin/compile/default-compile/inputFiles.lst:
```diff
@@ -0,0 +1,8 @@
+C:\Users\rando\Github\jpo-cvmanager\resources\keycloak\custom-user-provider\src\main\java\com\baeldung\auth\provider\mapper\CustomProtocolMapper.java
```
This seems to be referencing your directory structure - do we want to include that in the solution?

jacob6838 replied:
I removed all of the files in the target directory

payneBrandon commented on resources/keycloak/custom-user-provider/src/main/java/com/baeldung/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,93 @@
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+
+@JsonSerialize
+public class OrganizationObject {
```
Is this just a POJO/code representation of the relationship between organization and role for a particular user? Our SQL tables have this split out differently right?

jacob6838 replied:
Yes. The postgres table stores this in the user_organization table, which links a user to an organization and a role. This structure just holds this information in memory in the custom user provider, after multiple user_organization records are joined

payneBrandon commented on resources/keycloak/custom-user-provider/src/main/java/com/baeldung/auth/provider/user/CustomUserStorageProviderFactory.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public void onUpdate(KeycloakSession session, RealmModel realm, ComponentModel oldModel, ComponentModel newModel) {
+        log.debug("[I94] onUpdate()" );
```
Old log statements we want to remove?

jacob6838 replied:
I made them a little cleaner, i'm in favor of keeping them around since they only log in debug mode

payneBrandon replied:
Seems reasonable as long as we keep them generic to the implementation. I was very confused by the "[I94]" trying to reconcile a line number until I realized that must be your test data.

payneBrandon commented on resources/keycloak/custom-user-provider/src/main/java/com/baeldung/auth/provider/user/CustomUserStorageProviderFactory.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public void onCreate(KeycloakSession session, RealmModel realm, ComponentModel model) {
+        log.debug("[I99] onCreate()" );
```
Old statement to remove?

jacob6838 replied:
I made them a little cleaner, i'm in favor of keeping them around since they only log in debug mode

payneBrandon commented on resources/keycloak/custom-user-provider/src/main/java/com/baeldung/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    @Override
+    public UserAdapter getUserById(RealmModel realm, String id) {
```
I recognize this is an override so we can't easily rename it, but should we add some function comments specifying this is getting the user by keycloak id rather than user id?

jacob6838 replied:
I added some comments. I couldn't add anything to override the editor popup, because it's an overriden method

payneBrandon reviewed (changes requested):
A few questions so far. We also don't have any unit tests in here, can we get those added in as well?

jacob6838 committed:
Tweaking user registration and queries

Allowing users to register accounts through Google, and allowing the api to find users without organizations

jacob6838 committed:
Update middleware.py

jacob6838 committed:
Updating user provider comments for readability

jacob6838 committed:
Updating debug log statements

jacob6838 committed:
removing unused onUpdate and onCreate methods

jacob6838 committed:
Adding descriptive comments

jacob6838 committed:
Re-naming keycloak custom user provider java

jacob6838 committed:
Delete insert.sql

jacob6838 committed:
Updating keycloak docs for keystore generation

mcook42 commented on resources/keycloak/Dockerfile_localdev:
Question(blocking): The only differences I can find between this file and the Dockerfile are regarding the lines relating to the [keys used in the keytool command](https://github.com/CDOT-CV/jpo-cvmanager/pull/85/files#diff-d79113bd0a14d1c434a7d287339ddfc360ed7b27c3fe696d887b024a67d6497fR27-R36). By duplicating the majority of the Dockerfile in this new Dockerfile_localdev file, we're introducing a chance for drift between local and production docker images that may not be readily apparent. Is there a way to reduce duplication between this file and the original Dockerfile so that we only need to maintain one Dockerfile?

jacob6838 replied:
Good point - I re-combined the Dockerfiles, and added a dev script to generate a self-signed cert in resources/keycloak/create_local_cert.sh, and added documentation to the README.

mcook42 commented on resources/keycloak/custom-user-provider/.gitignore:
```diff
@@ -0,0 +1 @@
+target/**
```
praise: good usage of the `**` directive so that we don't need to keep adding specific `target/` directories to the gitignore should we need to add them in the future

mcook42 commented on resources/keycloak/custom-user-provider/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+FROM quay.io/keycloak/keycloak:latest as builder
```
Question(blocking): Have we considered pinning to a specific version of the image instead of using `latest`? Using latest can cause hard-to-detect and diagnose variability in builds [among other issues](https://vsupalov.com/docker-latest-tag/)

jacob6838 replied:
Good point, updated this to 21.1, the same version of keycloak we use in the rest of the cvmanager. I also created this spike work item to evaluate the newer version of keycloak, as they already have a version 26: https://dev.azure.com/SOC-OIT/CDOT/_workitems/edit/268677

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/mapper/CustomProtocolMapper.java:
```diff
@@ -0,0 +1,78 @@
+            transformAccessToken.getOtherClaims().put("cvmanager_data", UserObject.toTokenMap(user));
+        } catch (Exception e) {
+            log.error("Error transforming access token: " + e.getMessage());
+            e.printStackTrace();
```
```suggestion
            log.error("Error transforming access token: ", e);
```
chore: use standard error logging to put the necessary information in the log statement. Direct calls to `.printStackTrace()` are discouraged when writing code for production environments. The suggestion provided will still provide us with the error message and stack trace but allow us to leverage the logger more effectively for string concatenation and ensuring the stack trace is available in the log file (instead of stderr or other possible options).

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,324 @@
+                super_user
+            %s
+            ;
+                        """;
```
polish(non-blocking):  mark final field as final. use consistent indentation
```suggestion
    private static final String baseUserQuery = """
            SELECT
                keycloak_id,
                user_id,
                email,
                first_name,
                last_name,
                created_timestamp,
                super_user,
                COALESCE(
                    jsonb_agg(
                        jsonb_build_object('org', org_name, 'role', role)
                    ) FILTER (WHERE org_name IS NOT NULL AND role IS NOT NULL),
                    '[]'::jsonb
                ) AS organizations
            FROM (
                SELECT
                    users.keycloak_id,
                    users.user_id,
                    users.email,
                    users.first_name,
                    users.last_name,
                    users.created_timestamp,
                    users.super_user,
                    org.name AS org_name,
                    roles.name AS role
                FROM
                    public.users
                LEFT JOIN
                    public.user_organization AS uo ON uo.user_id = users.user_id
                LEFT JOIN
                    public.organizations AS org ON org.organization_id = uo.organization_id
                LEFT JOIN
                    public.roles ON roles.role_id = uo.role_id
            ) AS subquery
             %s
            GROUP BY
                user_id,
                keycloak_id,
                email,
                first_name,
                last_name,
                created_timestamp,
                super_user
            %s
            ;""";
```

jacob6838 replied:
Updated!

mcook42 commented on resources/keycloak/custom-user-provider/pom.xml:
```diff
@@ -0,0 +1,82 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.12.3</version> <!-- replace with the latest version -->
```
```suggestion
            <version>2.13.4.2</version>
```
The declared version has multiple vulnerabilities. If possible, we should use the most up-to-date version for this library to protect from said vulnerabilities. One vulnerability is https://osv.dev/vulnerability/GHSA-3x8x-79m2-3w2w

jacob6838 replied:
Are you using an extension to automatically detect these vulnerabilities?

mcook42 commented on resources/keycloak/custom-user-provider/pom.xml:
```diff
@@ -0,0 +1,82 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <java.version>17</java.version>
+
+        <keycloak.version>25.0.1</keycloak.version>
```
```suggestion
        <keycloak.version>25.0.6</keycloak.version>
```
If possible, we should upgrade to the 25.0.6 version to fix the vulnerabilities reported for version 25.0.1. See https://osv.dev/vulnerability/GHSA-j76j-rqwj-jmvv for one such vulnerability

jacob6838 replied:
Updated!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    @Override
+    public Stream<UserModel> searchForUserStream(RealmModel realm, String search, Integer firstResult,
```
Question (blocking): This method overrides a [deprecated method in the base class](https://github.com/keycloak/keycloak/blob/de973de80014f4dcaaaf021b9c7182834bb6d5d9/server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethodsProvider.java#L75). Should we override it if it's already deprecated?

jacob6838 replied:
removed!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/DbUtil.java:
```diff
@@ -0,0 +1,25 @@
+import org.keycloak.component.ComponentModel;
+import static com.cvmanager.auth.provider.user.CustomUserStorageProviderConstants.*;
+
+public class DbUtil {
```
Question: This question spans multiple files (`DbUtil`, `CustomUserStorageProvider`, and `CustomUserStorageProviderFactory`). Would it make more sense to move this single database connection method into the `CustomUserStorageProvider `or somewhere else? It seems like the intent of the `CustomUserStorageProvider` (see [close()](https://github.com/CDOT-CV/jpo-cvmanager/pull/85/files#diff-3337e5fbea3fb73084e2c44e9bd44d7b9a7e67eebec15478c21bb640dead1e4eR104)) is to encapsulate the database interactions with the User table. I see that we use the connection to validate the configuration in `CustomUserStorageProviderFactory`, but I'm not sure that warrants moving the connection management into a generic "util" class

jacob6838 replied:
Moved into CustomUserStorageProvider

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,324 @@
+
+    @Override
+    public Stream<UserModel> searchForUserByUserAttributeStream(RealmModel realm, String attrName, String attrValue) {
+        // TODO: Cast variables of certain types
```
Is this TODO still relevant?

jacob6838 replied:
nope, removed!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,324 @@
+            return users.stream();
+        } catch (SQLException ex) {
+            log.error("Database error in searchForUserByUserAttributeStream:", ex);
+            throw new RuntimeException("Database error:" + ex.getMessage(), ex);
```
question(non-blocking): to reduce duplication, should we consider creating something like a RuntimeStorageException that does what this line is doing? We have at least 8 occurrences of `RuntimeException("Database error:" + ex.getMessage(), ex)` in this file

jacob6838 replied:
Added a RuntimeStorageException class to reduce code duplication

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,324 @@
+        this.model = model;
+    }
+
+    private PreparedStatement getBaseUserQuery(Connection c, String where, String end) {
```
suggestion(non-blocking): rename this something like `prepareUserQuery` to accurately reflect what the method does.

jacob6838 replied:
renamed!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+        try {
+            return objectMapper.readValue(json, OrganizationObject.class);
+        } catch (JsonProcessingException e) {
+            log.error("Error parsing OrganizationObject from JSON: " + json, e);
```
```suggestion
            log.error("Error parsing OrganizationObject from JSON: {}", json, e);
```

jacob6838 replied:
updated!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+            return objectMapper.readValue(json,
+                    objectMapper.getTypeFactory().constructCollectionType(List.class, OrganizationObject.class));
+        } catch (JsonProcessingException e) {
+            log.error("Error parsing OrganizationObject from JSON: " + json, e);
```
```suggestion
            log.error("Error parsing OrganizationObject from JSON: {}", json, e);
```

jacob6838 replied:
updated!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+            return objectMapper.writeValueAsString(this);
+        } catch (JsonProcessingException e) {
+            log.error("Error serializing OrganizationObject to JSON", e);
+            return null;
```
```suggestion
            return "";
```

jacob6838 replied:
updated!

mcook42 commented on resources/keycloak/custom-user-provider/src/main/resources/ammplication.properties:
question: is `ammplication.properties` correct or is it a typo of `application.properties`?

jacob6838 replied:
yup... fixed

mcook42 reviewed (commented):
looking good. a few quesitons and suggestions

jacob6838 committed:
Setting keycloak version to 21.1

jacob6838 committed:
Updating keycloak dev certificate generation process

jacob6838 committed:
Update create_local_cert.sh

jacob6838 committed:
Updating package versions in custom-user-provider

jacob6838 committed:
Cleaning up logs/exceptions, re-naming for readability

jacob6838 committed:
Adding red highlight to admin rows without organizations

jacob6838 commented:
Just added a red highlight and tooltip to admin rows without any organizations linked. This assists in the case where a new user starts to join, but needs an admin to come in and add them to an organization
![image](https://github.com/user-attachments/assets/5ef5cf77-c91e-43fc-a632-a7656a227782)

mcook42 commented on resources/keycloak/create_local_cert.sh:
suggestion(non-blocking): run `git update-index --chmod=+x resources/keycloak/create_local_cert.sh` to mark this file as executable for everyone who clones this repo

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+COPY --from=builder /opt/keycloak/ /opt/keycloak/
+
+# change these values to point to a running postgres instance
+ENV KC_DB=postgres
+ENV KC_DB_URL=jdbc:postgresql://192.168.110.147/postgres?currentSchema=keycloak
+ENV KC_DB_USERNAME=postgres
+ENV KC_DB_PASSWORD=postgres
+ENV KC_HOSTNAME=localhost
+ENV KC_HOSTNAME=localhost
+ENV KC_LOG_LEVEL=DEBUG
```
suggestion(non-blocking): should we put these values in the `sample.env` file and read them in like we do [here](https://github.com/CDOT-CV/jpo-cvmanager/pull/85/files#diff-d79113bd0a14d1c434a7d287339ddfc360ed7b27c3fe696d887b024a67d6497fR29) so that we don't need to rely on users of this Dockerfile to read the comments?

jacob6838 replied:
This was only used for local dev, and made more sense when it was a separate project entirely. Now that it is in the cvmanager, it will only be run through the parent keycloak image. I chose to remove it entirely

jacob6838 committed:
Removing Dockerfile_localdev references

jacob6838 committed:
Adding KC_DB_URL_PROVIDER

jacob6838 committed:
removing KC_DB_URL_PROVIDER

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,315 @@
+        UserRegistrationProvider {
+
+    private static final Logger log = LoggerFactory.getLogger(CustomUserStorageProvider.class);
+    private KeycloakSession ksession;
+    private ComponentModel model;
```
suggestion(non-blocking):  By making fields final, you are reducing the number of "moving parts" readers of your code have to think about, making it much easier to reason about the code.
```suggestion
    private final KeycloakSession ksession;
    private final ComponentModel model;
```

jacob6838 replied:
done

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,315 @@
+        try (Connection c = getConnection(this.model)) {
+            PreparedStatement st = this.prepareUserQuery(c,
+                    String.format("WHERE %s = %s", attrName, attrValue),
+                    String.format("order by email"));
```
suggestion(non-blocking): since there are no format arguments, we can remove the call to `String.format`
```suggestion
                    "order by email");
```

jacob6838 replied:
done

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java:
```diff
@@ -0,0 +1,315 @@
+        }
+    }
+
+    public UserAdapter updateUser(RealmModel realm, UserObject user) {
```
question(non-blocking): does this method need to return `UserAdapter` or can it simply return `void`, since the only usage of this method (`save()`) doesn't use the return value

jacob6838 replied:
I like the paradigm of returning the updated value, even if it's not being used currently. It also enables checking the response for being null to know if the command errored out

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/UserAdapter.java:
```diff
@@ -0,0 +1,249 @@
+    public Map<String, List<String>> getAttributes() {
+        log.debug("getAttributes");
+        MultivaluedHashMap<String, String> attrs = new MultivaluedHashMap<>();
+        attrs.add(UserModel.USERNAME, entity.getEmail());
+        attrs.add(UserModel.EMAIL, entity.getEmail() != null ? entity.getEmail() : "");
```
question(non-blocking): why are we checking for a null email in the EMAIL case but not in the USERNAME case?

jacob6838 replied:
added checks in both locations

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/UserObject.java:
```diff
@@ -0,0 +1,147 @@
+
+@JsonSerialize
+public class UserObject {
+    private static final Logger log = LoggerFactory.getLogger(OrganizationObject.class);
```
```suggestion
    private static final Logger log = LoggerFactory.getLogger(UserObject.class);
```

jacob6838 replied:
derp

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/UserObject.java:
```diff
@@ -0,0 +1,147 @@
+
+    @JsonProperty("keycloak_id")
+    private String id;
+    @JsonProperty("user_id")
+    private Integer user_id;
```
suggestion(non-blocking): since we're using the `@JsonProperty` annotation to map the json name to the pojo field name, we can use java's naming convention (camel case) instead
```suggestion
    @JsonProperty("user_id")
    private Integer userId;
```

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/UserObject.java:
```diff
@@ -0,0 +1,147 @@
+    public static UserObject fromJoinedResultSet(ResultSet re) {
+        UserObject user = new UserObject();
+        try {
+            user.setId(re.getString("keycloak_id"));
+            user.setUserId(re.getInt("user_id"));
+            user.setEmail(re.getString("email"));
+            user.setFirstName(re.getString("first_name"));
+            user.setLastName(re.getString("last_name"));
+            user.setCreatedTimestamp(re.getLong("created_timestamp"));
+            user.setSuperUser(re.getInt("super_user"));
```
suggestion(non-blocking): these lines are duplicated in `fromResultSet` below. We may want to consider extracting the duplication to save time with updates/bug fixes in the future

jacob6838 replied:
I created a new Constants.java file containing all of these keys. I thought a bit about whether to integrate them into the queries in the CustomUserStorageProvider, but that would complicate the query logic a lot if we wanted to re-factor it, and clutter the code a lot. If Java had better inline string fomatting I would for sure do it. I will leave this up to your final say though if you have strong opinions

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/UserObject.java:
```diff
@@ -0,0 +1,147 @@
+        return user_id;
+    }
+
+    public void setUserId(Integer user_id) {
```
suggestion(non-blocking): the same as the field name suggestion above
```suggestion
    public void setUserId(Integer userId) {
```

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/UserAdapter.java:
```diff
@@ -0,0 +1,249 @@
+                        : Stream.empty();
+            default:
+                return super.getAttributeStream(name);
+        }
```
suggestion(non-blocking): this can be simplified by using the "enhanced switch statement", using `ofNullable` (which does what the same as what you're already doing, but in fewer characters), and merging the USERNAME and EMAIL cases since they're doing the same thing
```suggestion
        return switch (name) {
            case UserModel.USERNAME, UserModel.EMAIL -> Stream.ofNullable(entity.getEmail());
            case UserModel.FIRST_NAME -> Stream.ofNullable(entity.getFirstName());
            case UserModel.LAST_NAME -> Stream.ofNullable(entity.getLastName());
            case "created_timestamp" -> Stream.ofNullable(String.valueOf(entity.getCreatedTimestamp()));
            case "super_user" -> Stream.ofNullable(String.valueOf(entity.getSuperUser()));
            case "organizations" -> Stream.ofNullable(OrganizationObject.toStringList(entity.getOrganizations()));
            default -> super.getAttributeStream(name);
        };
```

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/UserAdapter.java:
```diff
@@ -0,0 +1,249 @@
+    @Override
+    public String getFirstAttribute(String name) {
+        log.debug("getFirstAttribute: " + name);
+        switch (name) {
+            case UserModel.USERNAME:
+                return entity.getEmail();
+            case UserModel.EMAIL:
+                return entity.getEmail();
+            case UserModel.FIRST_NAME:
+                return entity.getFirstName();
+            case UserModel.LAST_NAME:
+                return entity.getLastName();
+            case "created_timestamp":
+                return String.valueOf(entity.getCreatedTimestamp());
+            case "super_user":
+                return String.valueOf(entity.getSuperUser());
+            case "organizations":
+                return entity.getOrganizations() != null ? OrganizationObject.toStringList(entity.getOrganizations())
+                        : "[]";
+            default:
+                return super.getFirstAttribute(name);
+        }
```
suggestion(non-blocking): same as the `getAttributeStream` suggesiton below
```suggestion
        return switch (name) {
            case UserModel.USERNAME, UserModel.EMAIL -> entity.getEmail();
            case UserModel.FIRST_NAME -> entity.getFirstName();
            case UserModel.LAST_NAME -> entity.getLastName();
            case "created_timestamp" -> String.valueOf(entity.getCreatedTimestamp());
            case "super_user" -> String.valueOf(entity.getSuperUser());
            case "organizations" ->
                    entity.getOrganizations() != null ? OrganizationObject.toStringList(entity.getOrganizations())
                            : "[]";
            default -> super.getFirstAttribute(name);
        };
```

jacob6838 replied:
I like this syntax a LOT more, updated this and 3 other switch/case statements to match

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/UserAdapter.java:
```diff
@@ -0,0 +1,249 @@
+    protected UserObject entity;
+    protected String keycloakId;
+    final CustomUserStorageProvider provider;
+    final RealmModel realm;
```
realm is already provided by the base class
```suggestion
```

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/UserAdapter.java:
```diff
@@ -0,0 +1,249 @@
+            case UserModel.LAST_NAME:
+                entity.setLastName(value);
+                break;
+            case "created_timestamp":
+                entity.setCreatedTimestamp(value == null ? null : Long.valueOf(value));
+                break;
+            case "super_user":
+                entity.setSuperUser(value == null ? null : Integer.valueOf(value));
+                break;
+            case "organizations":
```
Suggestion (non-blocking): Should we consider making these strings into constants to reduce duplication (and usage of magic strings)?

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/UserAdapter.java:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    private void save() {
+        provider.updateUser(realm, entity);
+    }
```
issue: I understand the desire to encapsulate the "save" functionality into a function like this, but by not taking in parameters (like realm and entity) we're making it harder for future developers to understand what this method does. I like the clarity of `updateUser` already, so could we just use a direct call to `updateUser` instead? The method call indicates what's being updated by showing that we're passing in `realm` and `entity`, so users of the method don't need to wonder what's being updated or how.

jacob6838 replied:
Updated to be named saveUpdatedUserAttributes and take in the entity to save. I slightly prefer separating it, because it's duplicated and it's a nice abstraction in case we change the updateUser method later. I added a javadoc comment to saveUpdatedUserAttributes describing what it relies on for clarity

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+
+@JsonSerialize
+public class OrganizationObject {
+    /*
+     * This class is used to represent a user's organization and role association object from the user_organization postgresql table
+     */
```
To make it easier for developers to use this documentation, we should consider making it into a Javadoc comment so IDE users can see the comment more easily (see the screenshot for example) 
![image](https://github.com/user-attachments/assets/64a5d24a-e2fe-4a19-b098-d9f0c59a4dd4)
```suggestion
/**
 * OrganizationObject represents a user's organization and role association object from the user_organization postgresql table
 */
@JsonSerialize
public class OrganizationObject {
```

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+    public static List<OrganizationObject> listFromString(String json) {
+        try {
+            if (json == null || json.isEmpty()) {
+                return null;
```
to protect against `NullPointerExceptions`, we can pass back an empty list instead
```suggestion
                return Collections.emptyList();
```

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+                    objectMapper.getTypeFactory().constructCollectionType(List.class, OrganizationObject.class));
+        } catch (JsonProcessingException e) {
+            log.error("Error parsing OrganizationObject from JSON: {}", json, e);
+            return null;
```
to protect against NullPointerExceptions, we can pass back an empty list instead
```suggestion
            return Collections.emptyList();
```

jacob6838 replied:
updated

mcook42 commented on resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/pojos/OrganizationObject.java:
```diff
@@ -0,0 +1,96 @@
+            return objectMapper.convertValue(org, Map.class);
+        } catch (Exception e) {
+            log.error("Error converting OrganizationObject to Map", e);
+            return null;
```
to protect from `NullPointerExceptions` we should consider returning an empty map instead
```suggestion
            return Collections.emptyMap();
```

jacob6838 replied:
updated

mcook42 reviewed (changes requested):
Overall, the code looks like it works, but it isn't easy to provide a thorough review when there aren't any unit tests to show us what kinds of data are coming in and what data we're trying to produce. Could you add some unit tests that show examples of data we ingest and produce?

jacob6838 committed:
Adding string constants, cleaning up code

jacob6838 committed:
Re-writing switch/case and updating null returns

jacob6838 committed:
Adding JavaDoc comments and removing Dockerfile

payneBrandon reviewed (changes requested):
Looks like we're still awaiting some unit tests

jacob6838 committed:
Adding CustomProtocolMapper unit test

jacob6838 committed:
Adding unit tests

jacob6838 committed:
Adding example user query and response

jacob6838 commented:
@payneBrandon @mcook42 I just finished up the unit tests and added a sample query/response, this is ready for re-review!

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider

mcook42 commented on resources/keycloak/custom-user-provider/src/test/java/com/cvmanager/auth/provider/user/pojos/OrganizationObjectTest.java:
```diff
@@ -0,0 +1,133 @@
+public class OrganizationObjectTest {
+
+    @Test
+    private void listFromString() {
```
chore: [JUnit can not exectute `private` test methods](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-classes-and-methods). From JUnit documentation:

> It is generally recommended to omit the public modifier for test classes, test methods, and lifecycle methods unless there is a technical reason for doing so – for example, when a test class is extended by a test class in another package. Another technical reason for making classes and methods public is to simplify testing on the module path when using the Java Module System.

```suggestion
    void listFromString() {
```

jacob6838 replied:
Updated all tests!

mcook42 commented on resources/keycloak/custom-user-provider/src/test/java/com/cvmanager/auth/provider/user/pojos/OrganizationObjectTest.java:
```diff
@@ -0,0 +1,133 @@
+import java.util.List;
+import java.util.Map;
+
+public class OrganizationObjectTest {
```
Suggestion(non-blocking): Like the callout below, the recommendation from JUnit is to leave out the access modifiers in test classes and methods unless there is a specific reason for doing so. See https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-classes-and-methods

jacob6838 replied:
Removed all access modifiers

mcook42 reviewed (changes requested):
overall, the code looks good to me. There is one chore that needs to be done in all test files, then I'm good with hitting approve.

chore: remove all `private` access modifiers from all test files, if the method is annotated with `@Test` so that the tests are executed by JUnit. [JUnit can not execute private test methods](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-classes-and-methods). From JUnit documentation:

> It is generally recommended to omit the public modifier for test classes, test methods, and lifecycle methods unless there is a technical reason for doing so – for example, when a test class is extended by a test class in another package. Another technical reason for making classes and methods public is to simplify testing on the module path when using the Java Module System.

jacob6838 committed:
Making all unit tests public

jacob6838 committed:
removing access modifiers from tests

payneBrandon commented on resources/keycloak/custom-user-provider/README.md:
```diff
@@ -0,0 +1,98 @@
+
+This is a maven Spring-boot java project. This project generated a .jar file in ./target/keycloak-custom-providers-0.1.0-SNAPSHOT.jar. Please see the commands below for installing the project and generating the .jar file:
+
+### Installation
```
Can we add a section in here on running tests?

jacob6838 replied:
Added a section in the README and added running tests to the CI pipeline!

payneBrandon reviewed (commented):
Can we update the ci.yml to include the new java app and build/test it?

jacob6838 committed:
Run keycloak-user-provider mvn tests in ci

jacob6838 committed:
Changing CI java version

jacob6838 committed:
Updating java ci image

jacob6838 committed:
Update ci.yml

payneBrandon commented on resources/keycloak/create_local_cert.sh:
```diff
@@ -0,0 +1,12 @@
+mkdir -p ./ssl_cert
+
+# generate self-signed ssl key and cert
+openssl req -newkey rsa:2048 -nodes -keyout ./ssl_cert/key.pem -x509 -days 3650 -out ./ssl_cert/cert.pem 
```
Minor issue, we might want to check if openssl is installed or call it out as a dependency in the README

jacob6838 replied:
Just added this to the README

jacob6838 committed:
Documenting keyclaok ssl and openssl requirements

payneBrandon reviewed (approved):
lgtm!

mcook42 commented on .github/workflows/ci.yml:
```diff
@@ -76,6 +76,37 @@ jobs:
+          distribution: 'adopt'
+          java-version: '21'
+
+      - name: Build with Maven
+        run: |
+          cd $GITHUB_WORKSPACE/resources/keycloak/custom-user-provider
+          mvn clean install
+
+      - name: Run Tests with Maven
+        run: |
+          cd $GITHUB_WORKSPACE/resources/keycloak/custom-user-provider
+          mvn test
```
chore: `mvn clean install` already runs the `test` task by default, so the `Run Tests with Maven` will cause the tests to run twice
```suggestion
      - name: Build and Test with Maven
        run: |
          cd $GITHUB_WORKSPACE/resources/keycloak/custom-user-provider
          mvn clean install
```

jacob6838 replied:
Updated!

jacob6838 committed:
Updating CI to build and run tests in 1 step

jacob6838 committed:
Removing keycloak certificate requirement

jacob6838 committed:
Merge branch 'develop' into keycloak-user-provider
